Cypress installation and initial tests

gustavo-rss opened the pull request:
This PR contains:

- Cypress installation
- eslint installation
- eslint-plugin-cypress installation
 - Tests:
1.  It goes back to the customers list when clicking the 'Back' button *
2. It keeps the filters when coming back from the customer details view *

gustavo-rss committed:
chore: add eslint and cypress to the project

gustavo-rss committed:
test: add test "It keeps the filters when coming back from the customer details view"

gustavo-rss committed:
test: add test "It goes back to the customers list when clicking the 'Back' button *"

wlsf82 commented on cypress.config.js:
```diff
@@ -0,0 +1,9 @@
+
+module.exports = defineConfig({
+  e2e: {
+    setupNodeEvents(on, config) {
```
Visto que a função `setuNodeEvents` não está em uso, que tal removê-la por completo?

Código não usado gera complexidade desnecessária, e o segredo de um código com um bom design é a **simplicidade**.

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,31 @@
+
+  it("It goes back to the customers list when clicking the 'Back' button *", () => {
+    // Act
+    cy.get('button[aria-label*="View company:"]').first().click();
```
Uma opção mais simples, e mais legível, seria algo como o seguinte:

```js
cy.contains('button', 'View').click()
```

Perceba que não preciso do `.first()` visto que o comando `contains` só retorna o primeiro elemento encontrado.

A propósito, na linha abaixo você havia usado o `contains`. 🤷‍♂️

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,31 @@
+    cy.get('option[value="Large Enterprise"]')
+      .should("be.visible")
+      .and("be.selected");
+    cy.get('option[value="HR"]').should("be.visible").and("be.selected");
```
Faltou consistência.

Nas linhas 26-28 você quebrou linha, porém aqui não.

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,31 @@
+describe("EngageSphere", () => {
+  beforeEach(() => cy.visit("http://localhost:3000/"));
+  beforeEach(() => cy.contains("button", "Accept").click());
```
Por qual motivo você usou dois _hooks_ `beforeEach`?

Um só bastava.

Veja um exemplo:

```js
beforeEach(() => {
  cy.visit('/')
  cy.contains('button', 'Accept').click()
})
```

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,31 @@
+describe("EngageSphere", () => {
+  beforeEach(() => cy.visit("http://localhost:3000/"));
```
Não é recomendado deixar a URL _hardcoded_ no código.

Segue uma explicação detalhada abaixo.

# URL da aplicação hardcoded nos testes

É comum alunos/as escrevem scripts de testes com a URL da aplicação codificada diretamente no(s) arquivo(s) de testes (`cypress/e2e/gui/*.cy.js`), conforme exemplo abaixo:

```js
cy.visit('http://localhost:3000')
```

## Por que tal prática fere um bom design de testes?

Tal prática impossibilita a execução dos mesmos testes contra diferentes ambientes (ex.: local, homologação e produção) sem a necessidade da alteração da URL no arquivo dos testes, ou na duplicação de testes para diferentes ambientes.

## Sugestão com um melhor design

Em vez de codificar a URL da aplicação direto no arquivo de testes, é recomendado definir a `baseUrl` no arquivo de configurações (`cypress.config.js`).

Por exemplo:

```js
const { defineConfig } = require('cypress') 

module.exports = defineConfig({ 
  e2e: { 
    baseUrl: 'http://localhost:3000', 
  }, 
})
```

Dessa forma, é possível passar somente a URL relativa no arquivo de testes (ex.: `cy.visit('/')`)

Isso permite a mudança da URL através de argumentos de linha de comando, possibilitando a fácil execução dos mesmos testes contra diferentes ambientes, sem a necessidade de modificação nos testes propriamente ditos.

Exemplo: `npx cypress run --config 
baseUrl='https://engage-sphere.vercel.app'`

O exemplo acima sobrescreve a `baseUrl` padrão pela URL de produção, para execução dos mesmos testes contra o ambiente produtivo.

**Obs.:** Para que tal abordagem seja efetiva, os testes devem ser implementados de forma que possam ser executados contra diferentes ambientes, ou seja, deve-se pensar em variáveis de ambiente para credenciais de acesso de diferentes ambientes, massa de dados, etc.

wlsf82 commented on package.json:
```diff
@@ -17,7 +17,12 @@
-    "@babel/plugin-proposal-private-property-in-object": "^7.21.11"
+    "@babel/plugin-proposal-private-property-in-object": "^7.21.11",
+    "@eslint/js": "^9.37.0",
+    "cypress": "^14.5.4",
```
Algum motivo especial para não usar a versão 15?

wlsf82 reviewed (changes requested):
Parabéns pela implementacão @gustavo-rss! 👏🏻 

Deixei alguns comentários ao longo do seu PR, os quais creio que lhe ajudarão a melhorar o design dos testes.

A propósito, agora que o ESLint já está configurado, recomendo focar na escrita dos testes, que é onde aprenderá mais.

gustavo-rss committed:
refactor: improvements from PR review

gustavo-rss committed:
test: It renders the footer with the right text and links *
finished

gustavo-rss committed:
test: add tests
"It shows the cookie banner on first visit"
"It accepts the cookie banner on first visit"
"It renders the 'Hi there' greeting when no name is provided"
"It renders the 'Hi Joe' greeting when name is provided"

wlsf82 commented on .prettierrc:
```diff
@@ -0,0 +1,15 @@
+    "arrowParens": "always",
+    "proseWrap": "preserve",
+    "endOfLine": "lf"
+  }
```
É uma boa prática deixar uma linha em branco ao final de cada arquivo para evitar a exibição do ícone de `No newline at end of file` no GitHub.

```suggestion
  }

```

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,95 @@
+
+  it("It renders the footer with the right text and links", () => {
+    // Arrange
+    const footerContainer = 'footer[class^="Footer_container"]';
```
Não vejo a necessidade de ser tão específico.

Só a tag `footer` aqui já cumpriria o papel, pois a app inteira só tem um _Footer_.

E se esse for o caso, eu nem mesmo armazenaria tal valor em uma variável, pois no fim, o valor `footer` já é autoexplicativo.

O que acha?

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,95 @@
+    // Assert
+    cy.contains(footerContainer, "Copyright 2025 - Talking About Testing").should("be.visible");
+
+    cy.contains(`${footerContainer} a`, "Blog")
```
Daí, aqui podia ser `cy.contains("footer a", "Blog")`

wlsf82 replied:
Simples!

wlsf82 replied:
O mesmo vale para os próximos links

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,95 @@
+  it("It shows the cookie banner on first visit", () => {
+    // Arrange
+    cy.clearCookie("cookieConsent");
+    cy.reload();
```
Legal!

Outra alternativa seria um `describe` novo para os testes do _banner_ de consentimento de _cookies_, onde no `beforeEach` o `cookieConsent` não é definido.

Mas se preferir deixar assim, ao menos crie um `context` específico para ambos os testes.
Dessa forma, você não precisará repetir estes dois passos em ambos os testes.

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -0,0 +1,95 @@
+    cy.contains("button", "Accept").click();
+
+    // Assert
+    cy.get('[class^="CookieConsent_banner"]').should("not.exist");
```
Que tal também verificar que o _cookie_ foi definido corretamente?

Algo assim:

```js
cy.getCookie('cookieConsent').should('have.property', 'value', 'accepted')
```

A propósito, que tal criar um teste também para o cenário em que o _banner_ de consentimento de _cookies_ é rejeitado?

wlsf82 reviewed (changes requested):
Muito bem @gustavo-rss! 🎉 

Melhorou bastante, hein?

Deixei mais alguns comentários, os quais creio que melhorarão ainda mais sua habilidades em design de testes.

Parabéns! 👏🏻 👏🏻

gustavo-rss committed:
test: add new test "It declines the cookies banner on first visit"
small refactor to apply PR suggestions

gustavo-rss committed:
test: add test "It renders the header with a heading, theme's toggle, and a text input field"

gustavo-rss committed:
test: add test "It opens and closes the messenger"

gustavo-rss committed:
test: add test "It makes sure all messenger's fields are mandatory and the first one is focused"

gustavo-rss committed:
test: add test "It shows and hides a success message when successfully submitting the messenger form"

gustavo-rss committed:
test: add test "It clears all the messenger's form fields when filling them, closing the messenger, and opening it again"

gustavo-rss committed:
test: add test "It shows the Company name and Action columns and hides the ID, Industry, Number of Employees, and Size columns in a mobile viewport"

gustavo-rss committed:
test: add test "It returns the correct status and body structure"

Daniloeng commented on cypress.config.js:
```diff
@@ -0,0 +1,7 @@
+  e2e: {
+    baseUrl: "http://localhost:3000",
+  },
+});
```
Adicionar uma linha no final do arquivo é considereado boas práticas.

Daniloeng commented on cypress/support/commands.js:
mesmo que não for utilizar, talvez faça sentido remover os comentários default.

Daniloeng commented on cypress/support/e2e.js:
mesmo que não for alterar as configurações e2e, talvez faça sentido remover os comentários default.

Daniloeng commented on package.json:
Por padrão, talvez faça sentido adicionar um linha no final do arquivo.

wlsf82 commented on .prettierrc:
```diff
@@ -12,4 +12,5 @@
-  }
\ No newline at end of file
+  }
+  
```
```suggestion
  

```

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+
+  it("It renders the header with a heading, theme's toggle, and a text input field", () => {
+    // Assert
+    cy.get("[class^='Header_container']").should("be.visible").and("contain.text", "EngageSphere");
```
Uma versão simplificada seria a seguinte:

```suggestion
    cy.contains("h1", "EngageSphere").should("be.visible");
```

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+    cy.get("[class^='Header_container'] button")
+      .should("be.visible")
+      .and("have.attr", "aria-label", "theme light activated");
+    cy.get("[class^='Header_container'] input")
```
Essas três linhas poderiam ser simplificadas assim:

```js
cy.get("input[placeholder='E.g., John Doe']").should("be.visible");
```

O que acha?

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+    cy.get("[class^='Messenger_box']").should("not.exist");
+  });
+
+  it("It makes sure all messenger's fields are mandatory and the first one is focused", () => {
```
Onde estão as verificações de que os campos são obrigatórios?

Devias fazer algo assim:

```js
it('makes sure all fields are mandatory and the first one is focused', () => {
  cy.getByClassThatStartsWith('Messenger_form')
    .find('input[type="text"]')
    .should('be.focused')
    .and('have.attr', 'required')
  cy.getByClassThatStartsWith('Messenger_form')
    .find('input[type="email"]')
    .should('have.attr', 'required')
  cy.getByClassThatStartsWith('Messenger_form')
    .find('textarea')
    .should('have.attr', 'required')
})
```

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
-  it("It renders the 'Hi there' greeting when no name is provided", () => {
+  it("It declines the cookies banner on first visit", () => {
+    // Arrange
+    cy.get('[class^="CookieConsent_banner"]').should("be.visible");
```
Esta verificação não me parece necessária, visto que ela está implícita na existência do botão _Decline_.

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+  });
+});
+
+describe("EngageSphere API test", () => {
```
Que tal um arquivo para testes de GUI e outro para testes de API?

Dessa forma, ambos poderiam ser executados em paralelo.

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+});
+
+describe("EngageSphere API test", () => {
+  beforeEach(() => {});
```
O que faz esse `beforeEach` vazio aqui?

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+        expect(customer.size).to.be.a("string");
+        expect(customer.industry).to.be.a("string");
+
+        if (customer.contactInfo !== null) {
```
Uma versão simplificada seria a seguinte?

```suggestion
        if (customer.contactInfo) {
```

wlsf82 commented on cypress/e2e/EngageSphere.cy.js:
```diff
@@ -32,64 +32,268 @@ describe("EngageSphere", () => {
+          expect(customer.contactInfo).to.be.null;
+        }
+
+        if (customer.address !== null) {
```
Mesma coisa aqui.

```suggestion
        if (customer.address) {
```

wlsf82 reviewed (changes requested):
Muito bem, @gustavo-rss! 👏🏻 

Neste último _round_, deixei alguns poucos novos comentários, os quais serão os toques finais para o design dos testes ficar ótimo.

Espero que tenha aprendido bastante nessa jornada.
